src/selectors_vm/compiler.rs

```diff
@@ -23,46 +31,134 @@ pub struct CompiledAttributeExprOperand {
+                        }
+                        // NOTE: selector value can't be converted to the given encoding, so
+                        // it won't ever match.
+                        Err(_) => compile_expr!(@unmatchable),
+                    }
+                }
+                &OnTagNameExpr::NthChild(nth) => {
+                    compile_expr!(|state, _| state.counter.is_nth(nth))
+                }
+                &OnTagNameExpr::NthOfType(nth) => {
+                    *flags |= ProgramFlags::NTH_OF_TYPE;
+                    compile_expr!(|state, _| state.type_counter.expect("Counter for type required at this point").is_nth(nth))
+                }
+            }
+        );
+    }
+}
+
+impl Compilable for Expr<OnAttributesExpr> {
+    fn compile(
+        &self,
+        encoding: &'static Encoding,
+        ExprSet { attribute_exprs, .. }: &mut ExprSet,
+        _: &mut ProgramFlags,
+    ) {
+        let &Self { negation, ref simple_expr } = self;
+        curry_compile_expr_macro!(negation);
+
+        #[inline]
+        fn compile_literal(encoding: &'static Encoding, lit: &str) -> Result<Bytes<'static>, HasReplacementsError> {
+            Bytes::from_str_without_replacements(lit, encoding).map(Bytes::into_owned)
+        }
+
+        #[inline]
+        fn compile_literal_lowercase(encoding: &'static Encoding, lit: &str) -> Result<Bytes<'static>, HasReplacementsError> {
+            compile_literal(encoding, &lit.to_ascii_lowercase())
         }
+
+        attribute_exprs.push(
+            match simple_expr {
+                OnAttributesExpr::Id(id) =>
+                    compile_literal(encoding, id)
+                        .map_or_else::<CompiledAttributeExpr, _, _>(
+                            |_| compile_expr!(@unmatchable),
+                            |id| compile_expr!(|_, m| m.has_id(&id))
+                        ),
+
+                OnAttributesExpr::Class(class) =>
+                    compile_literal(encoding, class)
+                        .map_or_else::<CompiledAttributeExpr, _, _>(
+                            |_| compile_expr!(@unmatchable),
+                            |class| compile_expr!(|_, m| m.has_class(&class))
+                        ),
+
+                OnAttributesExpr::AttributeExists(name) =>
+                    compile_literal(encoding, name)
+                        .map_or_else::<CompiledAttributeExpr, _, _>(
+                            |_| compile_expr!(@unmatchable),
+                            |name| compile_expr!(|_, m| m.has_attribute(&name))
+                        ),
+
+                &OnAttributesExpr::AttributeExpr(AttributeExpr { ref name, ref value, case_sensitivity, operator }) => {
+                    #[inline]
+                    fn compile_operands(encoding: &'static Encoding, name: &str, value: &str) -> Result<(Bytes<'static>, Bytes<'static>), HasReplacementsError> {
```

Comment:
Get rid of nested functions - they make code extremely hard to read. Ideally, as a rule of thumb functions should have such a length that it should be possible to read its code on 13 inch display without scrolling.